### PR TITLE
identify nca errors

### PR DESCRIPTION
**Pull request by shireenf-ibm**
Signed-off-by: shireenf-ibm <shireenf@ibm.com>

Hi @adisos,
from my insights on this #13 issue, I concluded that running the nca cmd from baseline-rules-verifier, may return with 0 / 1 values (as nca’s exit code, `nca_run`).
baseline-rules-verifier, always “translates” `nca_run= 0` as a satisfied rule and `nca_run = 1 `as violated rule. however, in some **cases**  a query returning with result = 1 does not necessarily mean the rule is violated but may indicate that the query was not executed.

Such cases are: (**to my understanding**)
1. when the configs of the query are not comparable
2. when the forbids/permits config does not include any network-policies

So in order to realize if the returned `nca_run` (query result) is 1 because of one of the above cases, or because the rule is indeed violated , I have added a flag (`nca_err`) in the query answer (in nca code - [PR link](https://github.com/IBM/network-config-analyzer/pull/272)) which will return True if one of the cases above occurred.

In this commit:
when a "nca call" returns with the `nca_err` flag equal to `True`, it will add the query output to an `err_messages` variable and count it as an "query not executed" instead of violated rule

However, since nca is a library, this PR may be merged only after updating the nca version (?!) with the relevant PR, so it may be installed correctly with the `requirements.txt`

### CODE REVIEW

**Review by adisos — COMMENTED** [review]
Hi @shireenf-ibm , a few comments:

- [x] I think we should add a try/catch block for: `nca_run = nca.nca_main(nca_args)` , to catch exception raised by nca run. It's another case where the query did not run, and should be indicated as such in the rule results.
- [x] I think you can update `RuleResults` to capture the cases where the query did not run as another type of result. `rule_res` would not by of type bool, but rather enumerated (e.g. satisfied/violated/unknown -for the case where the rule cannot be checked). Also, `details` may be used to explain why a rule is not valid, or why it cannot be checked (can we use nca output to get the details for a query that did not run?)
- [x] Consider adding a test for this new rule result

**Comment by shireenf-ibm** [comment]
A note:
while running tests, I realized that while we support name-spaced rules in the "baseline-rules" which produce a calico GlobalNetworkPolicy with `namespaceSelector `for both egress and ingress. (i.e may contain `namespaceSelector `field in the policy's spec.)
 NCA does not support a namespaced calico GlobalNetworkPolicy .

For example, running the verifier with `--baseline baseline-rules/examples/deny_circle_blue.yaml`
raised an "nca" exception :
> namespaceSelector is not a valid entry in the specification of network policy spec (deny-circle-blue.yaml, line 5).

Should we open a new issue for nca to support `namespaceSelector `in the spec of calico policy?

File: src/baseline_verify.py [hunk 1]
```diff
@@ -71,13 +75,19 @@ def __init__(self, netpol_file, baseline_rules_file, repo):
...
+            sys.stdout = new_stdout = StringIO()
+            sys.stderr = new_stderr = StringIO()
 
-        nca_run = nca.nca_main(nca_args)
+            nca_run = nca.nca_main(nca_args)
+
+        except Exception as e:
+            # stop redirecting stderr to the buffers
+            sys.stderr = old_stderr
+            raise Exception(e) from None
```
Comment (adisos) [inline src/baseline_verify.py:90]: I think we should avoid raising an exception from baseline verifier if nca raised an exception. It's another case where the rule is not checked. The exception string may be used as details for why the rule was not checked.

File: src/baseline_verify.py [hunk 2]
```diff
@@ -26,24 +27,27 @@
...
+    satisfied = 0
+    violated = 1
+    unchecked = 2
+
+
 @dataclass
 class RuleResults:
     rule_name: str   # The name of the rule
-    satisfied: bool  # Whether the rule is valid
+    rule_status: RuleStatus  # Whether the rule is valid, violated or not checked
```
Comment (adisos) [inline src/baseline_verify.py:39]: Does details need to be updated differently when the rule is not checked?

**Review by adisos — COMMENTED** [review]
Hi @shireenf-ibm , added a few comments. 
Also please add some tests that demonstrate the output for various cases where rules cannot be checked (with and without cases of nca exceptions).

**Comment by adisos** [comment]
> A note: while running tests, I realized that while we support name-spaced rules in the "baseline-rules" which produce a calico GlobalNetworkPolicy with `namespaceSelector `for both egress and ingress. (i.e may contain `namespaceSelector `field in the policy's spec.) NCA does not support a namespaced calico GlobalNetworkPolicy .
> 
> For example, running the verifier with `--baseline baseline-rules/examples/deny_circle_blue.yaml` raised an "nca" exception :
> 
> > namespaceSelector is not a valid entry in the specification of network policy spec (deny-circle-blue.yaml, line 5).
> 
> Should we open a new issue for nca to support `namespaceSelector `in the spec of calico policy?

Yes, I think we should add support for this in nca.

**Comment by shireenf-ibm** [comment]
hi @adisos,
In the last commit , I have added 2 tests:
1. A test with `illegal-netpol.yaml` as netpols. This networkpolicy includes a syntax error which will cause nca to raise an exception thus the rule will not be checked (this test passes here as expected)
2. A test with `ingress-netpol.yaml` as netpols (since permits does not work with Ingress), nca will return with exit code 5 and the relevant message of `Ignoring PermitsQuery for ingress-netpol.yaml with Ingress only`.
However, this test still does not pass here (nca returns with 1 - rule violated) since we still didn't update nca (and its version) with the relevant PR.

a note :
I chose to add "ingress policy" as the example of unchecked rule, because the cases when permits/forbids query may not be executed are :
1. when the policies have different typologies (not comparable) , but in the verifier we send ns_list and pod_list from the same "repo" arg so I didn't find a way to send incomparable policies
2. when permits/forbids config is empty but since this policy is produced by baseline-rules , it doesn't allow "empty rule"
3. the config type is not relevant for the query -> the 2. test above (added)

**Review by adisos — COMMENTED** [review]
Hi @shireenf-ibm , looks good.
Let's comment-out the "ingress policy" test (since it is not representing a typical use case, I think it's ok if you just verified it locally with your nca version that returned a value greater than 3).

**Review by adisos — APPROVED** [review]

**Pull request merged**